File: identity/src/main/java/com/android/identity/ConnectionMethodNfc.java
package com.android.identity;

import static java.nio.charset.StandardCharsets.UTF_8;

import android.content.Context;
import android.nfc.NdefRecord;
import android.util.Log;
import android.util.Pair;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

import co.nstant.in.cbor.CborBuilder;
import co.nstant.in.cbor.builder.MapBuilder;
import co.nstant.in.cbor.model.Array;
import co.nstant.in.cbor.model.DataItem;
import co.nstant.in.cbor.model.Map;
import co.nstant.in.cbor.model.Number;

/**
 * Connection method for NFC.
 */
public class ConnectionMethodNfc extends ConnectionMethod {
    private static final String TAG = "ConnectionMethodNfc";
    private final long mCommandDataFieldMaxLength;
    private final long mResponseDataFieldMaxLength;

    static final int METHOD_TYPE = 1;
    static final int METHOD_MAX_VERSION = 1;
    private static final int OPTION_KEY_COMMAND_DATA_FIELD_MAX_LENGTH = 0;
    private static final int OPTION_KEY_RESPONSE_DATA_FIELD_MAX_LENGTH = 1;

    /**
     * Creates a new connection method for NFC.
     *
     * @param commandDataFieldMaxLength  the maximum length for the command data field.
     * @param responseDataFieldMaxLength the maximum length of the response data field.
     */
    public ConnectionMethodNfc(long commandDataFieldMaxLength,
                               long responseDataFieldMaxLength) {
        mCommandDataFieldMaxLength = commandDataFieldMaxLength;
        mResponseDataFieldMaxLength = responseDataFieldMaxLength;
    }

    /**
     * Gets the maximum length for the command data field.
     *
     * @return the value.
     */
    public long getCommandDataFieldMaxLength() {
        return mCommandDataFieldMaxLength;
    }

    /**
     * Gets the maximum length for the response data field.
     *
     * @return the value.
     */
    public long getResponseDataFieldMaxLength() {
        return mResponseDataFieldMaxLength;
    }

    public @Override
    @NonNull
    DataTransport createDataTransport(@NonNull Context context,
                                      @DataTransport.Role int role,
                                      @NonNull DataTransportOptions options) {
        DataTransportNfc t = new DataTransportNfc(context, role, this, options);
        // TODO: set mCommandDataFieldMaxLength and mResponseDataFieldMaxLength
        return t;
    }

    @Override
    public @NonNull
    String toString() {
        return "nfc:cmd_max_length=" + mCommandDataFieldMaxLength
                + ":resp_max_length=" + mResponseDataFieldMaxLength;
    }

    @Nullable
    static ConnectionMethodNfc fromDeviceEngagement(@NonNull DataItem cmDataItem) {
        if (!(cmDataItem instanceof co.nstant.in.cbor.model.Array)) {
            throw new IllegalArgumentException("Top-level CBOR is not an array");
        }
        List<DataItem> items = ((Array) cmDataItem).getDataItems();
        if (items.size() != 3) {
            throw new IllegalArgumentException("Expected array with 3 elements, got " + items.size());
        }
        if (!(items.get(0) instanceof Number) || !(items.get(1) instanceof Number)) {
            throw new IllegalArgumentException("First two items are not numbers");
        }
        long type = ((Number) items.get(0)).getValue().longValue();
        long version = ((Number) items.get(1)).getValue().longValue();
        if (!(items.get(2) instanceof co.nstant.in.cbor.model.Map)) {
            throw new IllegalArgumentException("Third item is not a map");
        }
        DataItem options = (Map) items.get(2);
        if (type != METHOD_TYPE) {
            Log.w(TAG, "Unexpected method type " + type);
            return null;
        }
        if (version > METHOD_MAX_VERSION) {
            Log.w(TAG, "Unsupported options version " + version);
            return null;
        }
        return new ConnectionMethodNfc(
                Util.cborMapExtractNumber(options, OPTION_KEY_COMMAND_DATA_FIELD_MAX_LENGTH),
                Util.cborMapExtractNumber(options, OPTION_KEY_RESPONSE_DATA_FIELD_MAX_LENGTH));
    }

    static @Nullable
    ConnectionMethod fromNdefRecord(@NonNull NdefRecord record) {
        ByteBuffer payload = ByteBuffer.wrap(record.getPayload()).order(ByteOrder.LITTLE_ENDIAN);
        int version = payload.get();
        if (version != 0x01) {
            Logger.w(TAG, "Expected version 0x01, found " + version);
            return null;
        }

        int cmdLen = payload.get() & 0xff;
        int cmdType = payload.get() & 0xff;
        if (cmdType != 0x01) {
            Logger.w(TAG, "expected type 0x01, found " + cmdType);
            return null;
        }
        if (cmdLen < 2 || cmdLen > 3) {
            Logger.w(TAG, "expected cmdLen in range 2-3, got " + cmdLen);
            return null;
        }

        int commandDataFieldMaxLength = 0;
        for (int n = 0; n < cmdLen - 1; n++) {
            commandDataFieldMaxLength *= 256;
            commandDataFieldMaxLength += payload.get() & 0xff;
        }

        int rspLen = payload.get() & 0xff;
        int rspType = payload.get() & 0xff;
        if (rspType != 0x02) {
            Logger.w(TAG, "expected type 0x02, found " + rspType);
            return null;
        }
        if (rspLen < 2 || rspLen > 4) {
            Logger.w(TAG, "expected rspLen in range 2-4, got " + rspLen);
            return null;
        }

        int responseDataFieldMaxLength = 0;
        for (int n = 0; n < rspLen - 1; n++) {
            responseDataFieldMaxLength *= 256;
            responseDataFieldMaxLength += payload.get() & 0xff;
        }

        return new ConnectionMethodNfc(commandDataFieldMaxLength, responseDataFieldMaxLength);
    }

    @NonNull
    @Override
    DataItem toDeviceEngagement() {
        MapBuilder<CborBuilder> builder = new CborBuilder().addMap();
        builder.put(OPTION_KEY_COMMAND_DATA_FIELD_MAX_LENGTH, mCommandDataFieldMaxLength);
        builder.put(OPTION_KEY_RESPONSE_DATA_FIELD_MAX_LENGTH, mResponseDataFieldMaxLength);
        return new CborBuilder()
                .addArray()
                .add(METHOD_TYPE)
                .add(METHOD_MAX_VERSION)
                .add(builder.end().build().get(0))
                .end()
                .build().get(0);
    }

    private static void encodeInt(int dataType, int value, ByteArrayOutputStream baos) {
        if (value < 0x100) {
            baos.write(0x02); // Length
            baos.write(dataType);
            baos.write(value & 0xff);
        } else if (value < 0x10000) {
            baos.write(0x03); // Length
            baos.write(dataType);
            baos.write(value / 0x100);
            baos.write(value & 0xff);
        } else {
            baos.write(0x04); // Length
            baos.write(dataType);
            baos.write(value / 0x10000);
            baos.write((value / 0x100) & 0xff);
            baos.write(value & 0xff);
        }
    }

    @Override @NonNull
    Pair<NdefRecord, byte[]> toNdefRecord() {
        byte[] carrierDataReference = "nfc".getBytes(UTF_8);

        // This is defined by ISO 18013-5 8.2.2.2 Alternative Carrier Record for device
        // retrieval using NFC.
        //
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        baos.write(0x01);  // Version
        encodeInt(0x01, (int) mCommandDataFieldMaxLength, baos);
        encodeInt(0x02, (int) mResponseDataFieldMaxLength, baos);
        byte[] oobData = baos.toByteArray();

        NdefRecord record = new NdefRecord((short) 0x02, // type = RFC 2046 (MIME)
                "iso.org:18013:nfc".getBytes(UTF_8),
                carrierDataReference,
                oobData);

        // From 7.1 Alternative Carrier Record
        //
        baos = new ByteArrayOutputStream();
        baos.write(0x01); // CPS: active
        baos.write(carrierDataReference.length); // Length of carrier data reference
        try {
            baos.write(carrierDataReference);
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
        baos.write(0x01); // Number of auxiliary references
        byte[] auxReference = "mdoc".getBytes(UTF_8);
        baos.write(auxReference.length);  // Length of auxiliary reference 0 data
        baos.write(auxReference, 0, auxReference.length);
        byte[] acRecordPayload = baos.toByteArray();

        return new Pair<>(record, acRecordPayload);
    }
}

File: multipaz/src/commonMain/kotlin/org/multipaz/mdoc/connectionmethod/ConnectionMethodNfc.kt
package org.multipaz.mdoc.connectionmethod

import org.multipaz.mdoc.transport.MdocTransport
import org.multipaz.nfc.NdefRecord
import org.multipaz.nfc.Nfc
import org.multipaz.util.Logger
import kotlinx.io.bytestring.buildByteString
import kotlinx.io.bytestring.encodeToByteString
import kotlinx.io.Buffer
import kotlinx.io.bytestring.ByteStringBuilder
import kotlinx.io.readByteArray
import kotlinx.io.readByteString
import kotlinx.io.write
import org.multipaz.cbor.Cbor
import org.multipaz.cbor.addCborMap
import org.multipaz.cbor.buildCborArray
import org.multipaz.cbor.Cbor.decode
import org.multipaz.cbor.Cbor.encode
import org.multipaz.cbor.CborArray
import org.multipaz.cbor.CborMap
import org.multipaz.util.ByteDataReader
import org.multipaz.util.appendByteString
import org.multipaz.util.appendUInt16
import org.multipaz.util.appendUInt8
/**
 * Connection method for NFC.
 *
 * @param commandDataFieldMaxLength  the maximum length for the command data field.
 * @param responseDataFieldMaxLength the maximum length of the response data field.
 */
class ConnectionMethodNfc(
    val commandDataFieldMaxLength: Long,
    val responseDataFieldMaxLength: Long
): ConnectionMethod() {
    override fun equals(other: Any?): Boolean {
        return other is ConnectionMethodNfc &&
                other.commandDataFieldMaxLength == commandDataFieldMaxLength &&
                other.responseDataFieldMaxLength == responseDataFieldMaxLength
    }

    override fun toString(): String =
        "nfc:cmd_max_length=$commandDataFieldMaxLength:resp_max_length=$responseDataFieldMaxLength"

    override fun toDeviceEngagement(): ByteArray {
        return Cbor.encode(
            buildCborArray {
                add(METHOD_TYPE)
                add(METHOD_MAX_VERSION)
                addCborMap {
                    put(OPTION_KEY_COMMAND_DATA_FIELD_MAX_LENGTH, commandDataFieldMaxLength)
                    put(OPTION_KEY_RESPONSE_DATA_FIELD_MAX_LENGTH, responseDataFieldMaxLength)
                }
            }
        )
    }

    private fun ByteStringBuilder.appendEncodedInt(dataType: UByte, value: Int) {
        if (value < 0x100) {
            appendUInt8(0x02) // Length
            appendUInt8(dataType)
            appendUInt8(value)
        } else if (value < 0x10000) {
            appendUInt8(0x03) // Length
            appendUInt8(dataType)
            appendUInt16(value)
        } else {
            appendUInt8(0x04) // Length
            appendUInt8(dataType)
            appendUInt8(value / 0x10000)
            appendUInt16(value.and(0xFFFF))
        }
    }

    override fun toNdefRecord(
        auxiliaryReferences: List<String>,
        role: MdocTransport.Role,
        skipUuids: Boolean
    ): Pair<NdefRecord, NdefRecord>? {
        val carrierDataReference = "nfc".encodeToByteString()

        // This is defined by ISO 18013-5 8.2.2.2 Alternative Carrier Record for device
        // retrieval using NFC.
        //
        val record = NdefRecord(
            NdefRecord.Tnf.EXTERNAL_TYPE,
            Nfc.EXTERNAL_TYPE_ISO_18013_5_NFC.encodeToByteString(),
            carrierDataReference,
            buildByteString {
                appendUInt8(0x01) // Version
                appendEncodedInt(DATA_TYPE_MAXIMUM_COMMAND_DATA_LENGTH, commandDataFieldMaxLength.toInt())
                appendEncodedInt(DATA_TYPE_MAXIMUM_RESPONSE_DATA_LENGTH, responseDataFieldMaxLength.toInt())
            }
        )

        // From NFC Forum Connection Handover v1.5 section 7.1 Alternative Carrier Record
        //
        check(auxiliaryReferences.size < 0x100)
        val acRecord = NdefRecord(
            tnf = NdefRecord.Tnf.WELL_KNOWN,
            type = Nfc.RTD_ALTERNATIVE_CARRIER,
            payload = buildByteString {
                appendUInt8(0x01) // CPS: active
                appendUInt8(carrierDataReference.size) // Length of carrier data reference
                appendByteString(carrierDataReference)
                appendUInt8(auxiliaryReferences.size) // Number of auxiliary references
                for (auxRef in auxiliaryReferences) {
                    // Each auxiliary reference consists of a single byte for the length and then as
                    // many bytes for the reference itself.
                    val auxRefUtf8 = auxRef.encodeToByteString()
                    check(auxRefUtf8.size < 0x100)
                    appendUInt8(auxRefUtf8.size)
                    appendByteString(auxRefUtf8)
                }
            }
        )
        return Pair(record, acRecord)
    }

    companion object {
        private const val TAG = "ConnectionMethodNfc"

        // Defined in ISO 18013-5 8.2.2.2 Alternative Carrier Record for device retrieval using NFC
        //
        private const val DATA_TYPE_MAXIMUM_COMMAND_DATA_LENGTH: UByte = 0x01u
        private const val DATA_TYPE_MAXIMUM_RESPONSE_DATA_LENGTH: UByte = 0x02u

        const val METHOD_TYPE = 1L
        const val METHOD_MAX_VERSION = 1L
        private const val OPTION_KEY_COMMAND_DATA_FIELD_MAX_LENGTH = 0L
        private const val OPTION_KEY_RESPONSE_DATA_FIELD_MAX_LENGTH = 1L

        internal fun fromDeviceEngagement(encodedDeviceRetrievalMethod: ByteArray): ConnectionMethodNfc? {
            val array = Cbor.decode(encodedDeviceRetrievalMethod)
            val type = array[0].asNumber
            val version = array[1].asNumber
            require(type == METHOD_TYPE)
            if (version > METHOD_MAX_VERSION) {
                return null
            }
            val map = array[2]
            return ConnectionMethodNfc(
                map[OPTION_KEY_COMMAND_DATA_FIELD_MAX_LENGTH].asNumber,
                map[OPTION_KEY_RESPONSE_DATA_FIELD_MAX_LENGTH].asNumber
            )
        }

        internal fun fromNdefRecord(
            record: NdefRecord,
            role: MdocTransport.Role,
        ): ConnectionMethodNfc? {
            with (ByteDataReader(record.payload)) {
                val version = getUInt8().toInt()
                if (version != 0x01) {
                    Logger.w(TAG, "Expected version 0x01, found $version")
                    return null
                }
                val cmdLen = getUInt8()
                val cmdType = getUInt8()
                if (cmdType != DATA_TYPE_MAXIMUM_COMMAND_DATA_LENGTH) {
                    Logger.w(TAG, "expected type 0x01, found $cmdType")
                    return null
                }
                if (cmdLen < 2u || cmdLen > 3u) {
                    Logger.w(TAG, "expected cmdLen in range 2-3, got $cmdLen")
                    return null
                }
                var commandDataFieldMaxLength = 0u
                for (n in 0u until cmdLen - 1u) {
                    commandDataFieldMaxLength *= 256u
                    commandDataFieldMaxLength += getUInt8()
                }
                val rspLen = getUInt8()
                val rspType = getUInt8()
                if (rspType != DATA_TYPE_MAXIMUM_RESPONSE_DATA_LENGTH) {
                    Logger.w(TAG, "expected type 0x02, found $rspType")
                    return null
                }
                if (rspLen < 2u || rspLen > 4u) {
                    Logger.w(TAG, "expected rspLen in range 2-4, got $rspLen")
                    return null
                }
                var responseDataFieldMaxLength = 0u
                for (n in 0u until rspLen - 1u) {
                    responseDataFieldMaxLength *= 256u
                    responseDataFieldMaxLength += getUInt8()
                }
                return ConnectionMethodNfc(
                    commandDataFieldMaxLength.toLong(),
                    responseDataFieldMaxLength.toLong()
                )
            }
        }
    }
}